sycl/include/CL/sycl/backend.hpp

```diff
@@ -32,7 +33,7 @@ namespace sycl {
 namespace detail {
 template <backend Backend, typename T> struct BackendInput {
   // TODO replace usage of interop with specializations.
-  using type = typename interop<Backend, T>::type;
+  typename interop<Backend, T>::type NativeHandle;
```

Comment:
[sycl/include/CL/sycl/backend/opencl.hpp#L71](https://github.com/smaslov-intel/llvm/blob/interop2020/sycl/include/CL/sycl/backend/opencl.hpp#L71) specializes `BackendInput` for `kernel_bundle` but not `interop`. I think that needs to be changed with this.

Reply:
I agree, otherwise the https://github.com/smaslov-intel/llvm/blob/163dfdb5564ba559f45ba8f81a2f2c43b6a86bc8/sycl/include/CL/sycl/backend.hpp#L227 won't compile. The same is true for other specializations of BackendInput, e.g. https://github.com/smaslov-intel/llvm/blob/163dfdb5564ba559f45ba8f81a2f2c43b6a86bc8/sycl/include/CL/sycl/backend/opencl.hpp#L81

Reply:
BTW, this change caused some tests to fail because input interop type now is not an equal to native handle, but a structure containing native handle (and possibly more):

```
[2021-09-03T07:04:16.524Z] /netbatch/donb2483175_00/runDir/jenkins-dir/workspace/SYCL_CI/intel/Lin/Basic_E2E/llvm-test-suite/SYCL/Basic/interop/traits.cpp:31:3: error: static_assert failed due to requirement 'std::is_same_v<sycl::detail::BackendInput<sycl::backend::opencl, sycl::device>, _cl_device_id *>'
```
Do you see any problem with that? Can you think of a way to have BackendInput either a structure or scalar type depending on the needs? I think in this case many of code reuse possible now in backend.*pp will not be possible and each backend would need to add its own specialization for everything always. So I'd like to avoid that.

Thoughts?

Reply:
BTW, this also means that what previously was (for example):

``` auto C = make_context<backend::opencl>(ocl_context)```

becomes

``` auto C = make_context<backend::opencl>({ocl_context})```

I don't think the SYCL2020 regulates this, but earlier seems prettier to me

Reply:
> I don't think the SYCL2020 regulates this

Actually, I've found that in C.6. Interoperability with the OpenCL API there is a table 185 that says

| SyclType | backend_input_t<backend::opencl, SyclType> | backend_return_t<backend::opencl, SyclType>| Description |
| --- | --- | --- | --- |
| platform | cl_platform_id | cl_platform_id | A SYCL platform object encapsulates an OpenCL platform ID. |
| device | cl_device_id | cl_device_id | A SYCL device object encapsulates an OpenCL device ID. |
| context | cl_context | cl_context | A SYCL context object encapsulates an OpenCL context object. |

etc

So, we can't wrap them into a structure

Reply:
I have redone changes, the backend_input_t continues to be a scalar type unless needed otherwise (currently only for context in L0 backend).